### PR TITLE
Homework 11 complete

### PR DESCRIPTION
**Pull request by asish918**
# Homework 11 Asish

### Topics Covered:
- JDBC Connector
- MySQL

### Drift Video Explanation
[Drift](https://app.drift.com/video/share/abznSw566mJPrMKQahvA5XWtRMwjn2Dcnsa4uHX32emag)

### Output Screenshots - 

Screenshot 2024-01-31 at 23 04 25

Screenshot 2024-01-31 at 23 04 10

Screenshot 2024-01-31 at 23 04 00

Screenshot 2024-01-31 at 23 03 39

Screenshot 2024-01-31 at 23 03 33

Check the resources/assignment.http file for all the REST API calls.

### CODE REVIEW

**Comment by asish918** [comment]
| Metric | Value |
| ------ | ----- |
| alert_status | <b>:white_check_mark: OK</b> |
| bugs | 0 |
| code_smells | 0 |
| reliability_rating | A (0 Bugs)<br/>Note: A being best, E being worst |
| security_rating | A (0 Vulnerabilities)<br/>Note: A being best, E being worst |
| vulnerabilities | 0 |

File: backend/homework/11/src/main/java/asish/kdu/sring_jdbc_homework/controller/AppController.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    @Autowired
+    public AppController(AppService appService, UserService userService) {
+        this.appService = appService;
+        this.userService = userService;
+    }
+
+    @GetMapping("/test")
+    public ResponseEntity<String> testEndpoint() {
+        log.info("TEST GET Request successful");
+        return new ResponseEntity<>("Server working. Test successful....", HttpStatus.OK);
```
Comment (rishav-kd) [inline backend/homework/11/src/main/java/asish/kdu/sring_jdbc_homework/controller/AppController.java:28]: lets move this to service layers

File: backend/homework/11/src/main/java/asish/kdu/sring_jdbc_homework/dtos/ShiftTypeDTO.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+@Data
+@AllArgsConstructor
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class ShiftTypeDTO {
+    private String uqName;
```
Comment (rishav-kd) [inline backend/homework/11/src/main/java/asish/kdu/sring_jdbc_homework/dtos/ShiftTypeDTO.java:11]: lets use `@JsonProperty` here too.

File: backend/homework/11/src/main/java/asish/kdu/sring_jdbc_homework/dtos/UsersDTO.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+@Data
+@AllArgsConstructor
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class UsersDTO {
+    private String id;
+    private String username;
```
Comment (rishav-kd) [inline backend/homework/11/src/main/java/asish/kdu/sring_jdbc_homework/dtos/UsersDTO.java:12]: lets use @Valid here

File: backend/homework/11/src/main/java/asish/kdu/sring_jdbc_homework/dtos/mappers/CompleteMapper.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+import asish.kdu.sring_jdbc_homework.dtos.*;
+
+public class CompleteMapper {
+    private CompleteMapper() {
+
+    }
+
+    public static UsersDTO toUsersDto(CompleteDTO completeDTO, String tenantId, String userId) {
+        return new UsersDTO(
```
Comment (rishav-kd) [inline backend/homework/11/src/main/java/asish/kdu/sring_jdbc_homework/dtos/mappers/CompleteMapper.java:11]: Lets use `builder().build()` here.

File: backend/homework/11/src/main/java/asish/kdu/sring_jdbc_homework/exceptions/GlobalExceptions.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
+@Slf4j
+public class GlobalExceptions {
+    @ExceptionHandler({NoSQLDataFound.class})
+    public ResponseEntity<String> noData(NoSQLDataFound e){
+        log.error("No data found for the query.");
```
Comment (rishav-kd) [inline backend/homework/11/src/main/java/asish/kdu/sring_jdbc_homework/exceptions/GlobalExceptions.java:15]: Lets create custom message strings for these.

File: backend/homework/11/src/main/java/asish/kdu/sring_jdbc_homework/repository/ShiftRepository.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    private JdbcTemplate jdbcTemplate;
+
+    @Autowired
+    public ShiftRepository(DataSource dataSource) {
+        this.jdbcTemplate = new JdbcTemplate(dataSource);
+    }
+
+    @SuppressWarnings("java:S2077")
+    public void createShiftType(ShiftType shiftType) {
+        String sql = "INSERT INTO shift_types (uq_name, description, active, created_by, updated_by, time_zone, tenant_id) VALUES (?, ?, ?, ?, ?, ?, UUID_TO_BIN(?))";
```
Comment (rishav-kd) [inline backend/homework/11/src/main/java/asish/kdu/sring_jdbc_homework/repository/ShiftRepository.java:24]: lets separate this out as a constant

File: backend/homework/11/src/main/java/asish/kdu/sring_jdbc_homework/services/UserService.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+@Service
+public class UserService {
+    private UserRepository userRepository;
+
+    @Autowired
+    public UserService(UserRepository userRepository) {
+        this.userRepository = userRepository;
+    }
+
+    public void addUser(UsersDTO usersDTO) {
```
Comment (rishav-kd) [inline backend/homework/11/src/main/java/asish/kdu/sring_jdbc_homework/services/UserService.java:22]: Lets add validation here.

**Review by rishav-kd — COMMENTED** [review]
Looks good. Focus on comments

**Pull request closed without merging**